File: src/my_ros_tutorial_pkgs/my_ros_tutorial_py/py_pubsub/py_pubsub/publisher_member_function.py
import rclpy
from rclpy.node import Node

from std_msgs.msg import String


class MinimalPublisher(Node):
    def __init__(self):
        super().__init__('minimal_publisher')

        self.publisher_ = self.create_publisher(String, 'topic', 10)
        self.timer_ = self.create_timer(0.5, self.timer_callback)
        
        self.count = 0
    
    def timer_callback(self):
        self.msg = String()

        self.msg.data = f"hello world: {self.count}"
        self.count += 1

        self.get_logger().info(f'publishing "{self.msg.data}"')
        
        self.publisher_.publish(self.msg)


def main(args=None):
    rclpy.init(args=args)
    node = MinimalPublisher()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()


if __name__ == "__main__":
    main()

File: src/my_ros_tutorial_pkgs/my_ros_tutorial_cpp/cpp_pubsub/src/publisher_member_function.cpp
#include <chrono>
#include <functional>

#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"


using namespace std::literals::chrono_literals;

// Create node
class MinimalPublisher: public rclcpp::Node
{
public:
  // Constructor
  MinimalPublisher()
  : Node("minimal_publisher"), count(0) // Initializer list
  {
    publisher_ = this->create_publisher<std_msgs::msg::String>("topic", 10); // Create publisher
    
    // Create timer
    timer_ = this->create_wall_timer(
      500ms,
      [this]() { timer_callback(); } // use lambda instead of binder
    );
  }

private:
  // Create callback for timer
  void timer_callback() {
    // Create and populate msg
    std_msgs::msg::String msg;
    msg.data = "Hello World: " + std::to_string(count++);

    publisher_->publish(msg); // Publish msg

    RCLCPP_INFO(this->get_logger(), "Publishing: \"%s\"", msg.data.c_str());
  }

  rclcpp::Publisher<std_msgs::msg::String>::SharedPtr publisher_;
  rclcpp::TimerBase::SharedPtr timer_;
  
  int count;
};


int main(int argc, char* argv[]) {
  rclcpp::init(argc, argv); // Initialize ROS2 communication
  std::shared_ptr<MinimalPublisher> node = std::make_shared<MinimalPublisher>(); // Create node
  rclcpp::spin(node); // Spin node
  rclcpp::shutdown(); // Shutdown ROS2 communication

  return 0;
}